### PR TITLE
Make controlling framebuffer configurable and flexible

### PR DESCRIPTION
**Pull request by WayneLin1992**
In the original version, vcam_in_queue_setup() called
vmalloc twice for frame buffers, and we can't know what
size are buffers.This patch main includes
register_framebuffer() and replaces vcam_in_queue_setup()
in vcam initiated. It defaults frame buffer 640 * 480
and resizes the frame buffer when scaling ON.

To check the frame buffer information, use the following command:
$ sudo fbset -fb /dev/fb1 --info

```
mode "640x480"
    geometry 640 480 640 480 24
    timings 0 0 0 0 0 0 0
    rgba 8/0,8/8,8/16,0/0
endmode

Frame buffer device information:
    Name        : vcamfb
    Address     : 0xffffb5a6427d1000
    Size        : 921600
    Type        : PACKED PIXELS
    Visual      : TRUECOLOR
    XPanStep    : 1
    YPanStep    : 1
    YWrapStep   : 1
    LineLength  : 1920
    Accelerator : No
```
after the scaling ON
```
mode "1280x720"
    geometry 1280 720 1280 720 24
    timings 0 0 0 0 0 0 0
    rgba 8/0,8/8,8/16,0/0
endmode

Frame buffer device information:
    Name        : vcamfb
    Address     : 0xffffb5a646985000
    Size        : 2764800
    Type        : PACKED PIXELS
    Visual      : TRUECOLOR
    XPanStep    : 1
    YPanStep    : 1
    YWrapStep   : 1
    LineLength  : 3840
    Accelerator : No
```

### CODE REVIEW

File: device.c [hunk 1]
```diff
@@ -186,7 +187,19 @@ static int vcam_try_fmt_vid_cap(struct file *file,
         }
         dev->output_format.sizeimage =
             dev->output_format.bytesperline * dev->output_format.height;
+
+        /* resize the videobuffer */
         vcam_update_format_cap(dev, false);
+
+        /* resize the framebuffer */
+        destroy_framebuffer((const char *) dev->vcam_fb_fname);
```
Comment (jserv) [inline device.c:195]: `destroy_framebuffer` is considered harmful. It might be associated with different proc entries after new `init_framebuffer` requst.

Comment (jserv) [reply device.c:195]: Check this project: https://github.com/hadi77ir/linux-module-virtfb
You can use `fbset` to change its mode/resolution on the fly.

**Review by jserv — CHANGES_REQUESTED** [review]
The subject was misleading. You should emphasize on the motivation and reasons why you are going to change.

File: fb.c [hunk 2]
```diff
@@ -145,3 +145,148 @@ void destroy_framebuffer(const char *proc_fname)
...
+    return 0;
+}
+
+static const struct fb_ops vcamfb_ops = {
+    .fb_set_par = vcam_fb_set_par,
+    .fb_check_var = vcam_fb_check_var,
+};
+
+static struct fb_fix_screeninfo vfb_fix = {
+    .id = "vcam fb",
```
Comment (jserv) [inline fb.c:202]: Why did you call it `vcam fb` instead of `vcamfb`?

File: fb.c [hunk 3]
```diff
@@ -145,3 +145,148 @@ void destroy_framebuffer(const char *proc_fname)
...
+    .type = FB_TYPE_PACKED_PIXELS,
+    .visual = FB_VISUAL_TRUECOLOR,
+    .xpanstep = 1,
+    .ypanstep = 1,
+    .ywrapstep = 1,
+    .accel = FB_ACCEL_NONE,
+};
+
+static const struct fb_videomode vfb_default = {
+    .xres = 640,
```
Comment (jserv) [inline fb.c:212]: Can you share the default size with V4L2 part? That is, we only have to maintain one set of default resolution.

File: fb.h [hunk 4]
```diff
@@ -7,4 +7,9 @@ struct proc_dir_entry *init_framebuffer(const char *proc_fname,
                                         struct vcam_device *dev);
 void destroy_framebuffer(const char *proc_fname);
 
+int init_vcamfb(struct vcam_device *dev);
+
+void destroy_vcamfb(struct vcam_device *dev);
+
+void vcam_update_vcamfb(struct vcam_device *dev);
```
Comment (jserv) [inline fb.h:14]: Append one blank line here.

File: fb.c [hunk 5]
```diff
@@ -145,3 +145,148 @@ void destroy_framebuffer(const char *proc_fname)
...
+    if (var->yres > var->yres_virtual)
+        var->yres_virtual = var->yres;
+
+    var->xres_virtual = var->xoffset + var->xres;
+    var->yres_virtual = var->yoffset + var->yres;
+
+    var->bits_per_pixel = 24;
+
+    switch (var->bits_per_pixel) {
+    case 24:
```
Comment (jserv) [inline fb.c:167]: Only bpp = 24 is supported?

Comment (WayneLin1992) [reply fb.c:193]: It is the screeninfo variable, so It can support rgb type for example rgb565, rgb5551.

File: fb.c [hunk 6]
```diff
@@ -145,3 +145,179 @@ void destroy_framebuffer(const char *proc_fname)
...
+    if (var->bits_per_pixel <= 16)
+        var->bits_per_pixel = 16;
+    else if (var->bits_per_pixel <= 24)
+        var->bits_per_pixel = 24;
+    else
+        return -EINVAL;
+
+    switch (var->bits_per_pixel) {
+    case 16:
+        if (var->transp.length) {
```
Comment (jserv) [inline fb.c:173]: Add comments which denotes the color space.

**Review by jserv — CHANGES_REQUESTED** [review]
Improve the descriptions of git commit message.

The statement "Registers framebuffer easy to trace its format" was meaningless to me.

**Comment by jserv** [comment]
> Improve the descriptions of git commit message.
> The statement "Registers framebuffer easy to trace its format" was meaningless to me.

In addition, improve the subject of this pull request.

**Review by jserv — CHANGES_REQUESTED** [review]
The subject of this pull request and git commit is a bit redundant. You can say,
> Allow flexible configurations for controlling framebuffer

File: fb.c [hunk 7]
```diff
@@ -145,3 +145,288 @@ void destroy_framebuffer(const char *proc_fname)
...
+        var->yres_virtual = var->yres;
+
+    var->xres_virtual = var->xoffset + var->xres;
+    var->yres_virtual = var->yoffset + var->yres;
+
+    if (var->bits_per_pixel <= 16)
+        var->bits_per_pixel = 16;
+    else if (var->bits_per_pixel <= 24)
+        var->bits_per_pixel = 24;
+    else
```
Comment (jserv) [inline fb.c:168]: Is it possible to support bpp = 32 ?

File: fb.c [hunk 8]
```diff
@@ -145,3 +145,290 @@ void destroy_framebuffer(const char *proc_fname)
...
+        var->yres = 1;
+    if (var->xres > var->xres_virtual)
+        var->xres_virtual = var->xres;
+    if (var->yres > var->yres_virtual)
+        var->yres_virtual = var->yres;
+
+    var->xres_virtual = var->xoffset + var->xres;
+    var->yres_virtual = var->yoffset + var->yres;
+
+    if (var->bits_per_pixel <= 16)
```
Comment (jserv) [inline fb.c:164]: I dislike the way you set bpp. Can you use the trick of align up as Linux kernel?
See https://lists.kernelnewbies.org/pipermail/kernelnewbies/2013-November/009229.html

File: fb.c [hunk 9]
```diff
@@ -161,12 +161,13 @@ static int vcam_fb_check_var(struct fb_var_screeninfo *var,
     var->xres_virtual = var->xoffset + var->xres;
     var->yres_virtual = var->yoffset + var->yres;
 
+    /* check bpp */
     if (var->bits_per_pixel <= 16)
-        var->bits_per_pixel = 16;
+       var->bits_per_pixel =  ALIGN(16, 4);
```
Comment (jserv) [inline fb.c:166]: No, you can avoid the repeated if-else statements by `ALIGN` macro.

Comment (WayneLin1992) [reply fb.c:166]: the bpp = 24 is not the power of two.
I thought that write on below.
```cpp
int bpp = var->bits_per_pixel;
if (bpp > 16 && bpp <= 24)
    var->bits_per_pixel = 24;
else
    var->bits_per_pixel = ALIGN(bpp, 4);
```

File: fb.c [hunk 10]
```diff
@@ -145,3 +145,289 @@ void destroy_framebuffer(const char *proc_fname)
...
+    if (var->xres > var->xres_virtual)
+        var->xres_virtual = var->xres;
+    if (var->yres > var->yres_virtual)
+        var->yres_virtual = var->yres;
+
+    var->xres_virtual = var->xoffset + var->xres;
+    var->yres_virtual = var->yoffset + var->yres;
+
+    /* check bpp */
+    bpp = var->bits_per_pixel;
```
Comment (jserv) [inline fb.c:166]: You shall raise an error if bpp is invalid.

**Comment by WayneLin1992** [comment]
`ALIGN(17, 8)` = 24 & ~7 = 11000 & 11000 = 11000
`ALIGN(10, 8)` = 17 & ~7 = 10001 & 11000 = 10000

File: fb.c [hunk 11]
```diff
@@ -145,3 +145,289 @@ void destroy_framebuffer(const char *proc_fname)
...
+        var->yres_virtual = var->yres;
+
+    var->xres_virtual = var->xoffset + var->xres;
+    var->yres_virtual = var->yoffset + var->yres;
+
+    /* check bpp value ALIGN 8 */
+    bpp = var->bits_per_pixel;
+    if (bpp <= 8 || bpp > 32)
+        return -EINVAL;
+    else
```
Comment (jserv) [inline fb.c:169]: `else` is not necessary.

File: fb.c [hunk 12]
```diff
@@ -145,3 +145,289 @@ void destroy_framebuffer(const char *proc_fname)
...
+        u32 color;
+
+        if (regno >= 16)
+            return -EINVAL;
+
+        color = (red << info->var.blue.offset) |
+                (green << info->var.green.offset) |
+                (blue << info->var.red.offset) |
+                (transp << info->var.transp.offset);
+        switch (info->var.bits_per_pixel) {
```
Comment (jserv) [inline fb.c:276]: Why did you need this switch-case?

File: fb.c [hunk 13]
```diff
@@ -145,3 +145,281 @@ void destroy_framebuffer(const char *proc_fname)
...
+    int retval = -ENOMEM;
+    if (!(videomemory = vmalloc(size)))
+        return retval;
+    dev->buffer_addr = videomemory;
+    dev->offset_buffer = dev->input_format.sizeimage;
+    q->buffers[0].data = videomemory;
+    q->buffers[0].filled = 0;
+    q->buffers[1].data = (void *) (videomemory + dev->offset_buffer);
+    q->buffers[1].filled = 0;
+    memset(&q->dummy, 0x00, sizeof(struct vcam_in_buffer));
```
Comment (jserv) [inline fb.c:327]: Use `0` rather than `0x00` for simplicity.

File: fb.c [hunk 14]
```diff
@@ -145,3 +145,281 @@ void destroy_framebuffer(const char *proc_fname)
...
+        unsigned int size;
+        vfree(dev->buffer_addr);
+        dev->offset_buffer = dev->input_format.sizeimage;
+        size = dev->input_format.sizeimage * 2;
+        dev->buffer_addr = vmalloc(size);
+        q->buffers[0].data = dev->buffer_addr;
+        q->buffers[1].data = (void *) (dev->buffer_addr + dev->offset_buffer);
+        q->buffers[0].filled = 0;
+        q->buffers[1].filled = 0;
+        memset(&q->dummy, 0x00, sizeof(struct vcam_in_buffer));
```
Comment (jserv) [inline fb.c:404]: Ditto.

File: fb.c [hunk 15]
```diff
@@ -145,3 +145,281 @@ void destroy_framebuffer(const char *proc_fname)
...
+    var->blue.msb_right = 0;
+    var->transp.msb_right = 0;
+
+    return 0;
+}
+
+static int vcam_fb_set_par(struct fb_info *info)
+{
+    switch (info->var.bits_per_pixel) {
+    case 16:
```
Comment (jserv) [inline fb.c:230]: Why did you use switch-case here?

File: fb.c [hunk 16]
```diff
@@ -145,3 +145,281 @@ void destroy_framebuffer(const char *proc_fname)
...
+        info->fix.visual = FB_VISUAL_TRUECOLOR;
+        break;
+    }
+    return 0;
+}
+
+static int vcam_fb_mmap(struct fb_info *info, struct vm_area_struct *vma)
+{
+    int ret;
+    ret =
```
Comment (jserv) [inline fb.c:246]: Rewrite as:
```c
int ret = remap_vmalloc_range(vma, (void *) info->fix.smem_start, vma->vm_pgoff);
```

**Comment by WayneLin1992** [comment]
After this pull request, should I write this on `README.md`.

**Comment by jserv** [comment]
> After this pull request, should I write this on `README.md`.

You should update `README.md` within this pull request.

File: device.h [hunk 17]
```diff
@@ -71,6 +72,10 @@ struct vcam_device {
     char vcam_fb_fname[FB_NAME_MAXLENGTH];
     struct proc_dir_entry *vcam_fb_procf;
     struct mutex vcam_mutex;
+    /* framebuffer info  */
+    struct fb_info info;
+    void *buffer_addr;
```
Comment (jserv) [inline device.h:77]: It is a bit confusing between `buffer_addr` and `offset_buffer`. Can you clarify? That is, `fb_addr` and `fb_offset`.

File: README.md [hunk 18]
```diff
@@ -35,9 +35,10 @@ The module can be loaded to Linux kernel by runnning the command:
 $ sudo insmod vcam.ko
 ```
 
-Expectedly, two device nodes will be created in `/dev`:
+Expectedly, three device nodes will be created in `/dev`:
 * videoX - V4L2 device;
 * vcamctl - Control device for virtual camera(s), used by control utility `vcam-util`;
+* fbX - frame buffer device;
```
Comment (jserv) [inline README.md:41]: Change the description to
> controlling framebuffer device

No space is separated between "frame" and "buffer."

Comment (jserv) [reply README.md:41]: See https://github.com/dvdhrm/docs/blob/master/drm-howto/modeset.c

File: device.h [hunk 19]
```diff
@@ -71,6 +72,10 @@ struct vcam_device {
     char vcam_fb_fname[FB_NAME_MAXLENGTH];
     struct proc_dir_entry *vcam_fb_procf;
     struct mutex vcam_mutex;
+    /* framebuffer info  */
+    struct fb_info info;
```
Comment (jserv) [inline device.h:76]: The name `info` might be too generic to distinguish. Clarify it as well.

Comment (jserv) [reply device.h:76]: Alternatively, you can replace the three members into `void *fb_priv`, which encapsulates framebuffer specific data. And, you don't have to include `<linux/fb.h>` in header `device.h`.

File: fb.c [hunk 20]
```diff
@@ -1,12 +1,19 @@
...
+#include <linux/fb.h>
 #include <linux/proc_fs.h>
 #include <linux/spinlock.h>
 #include <linux/version.h>
 
 #include "fb.h"
 #include "videobuf.h"
 
+struct vcamfb_info {
+    struct fb_info info;
```
Comment (jserv) [inline fb.c:12]: Since the members are now in `struct vcamfb_info`, the prefix `fb_` is not necessary. You can safely rename them into `addr` and `offset`.

File: README.md [hunk 21]
```diff
@@ -35,9 +35,10 @@ The module can be loaded to Linux kernel by runnning the command:
 $ sudo insmod vcam.ko
 ```
 
-Expectedly, two device nodes will be created in `/dev`:
+Expectedly, three device nodes will be created in `/dev`:
 * videoX - V4L2 device;
 * vcamctl - Control device for virtual camera(s), used by control utility `vcam-util`;
+* fbX - controlling framebuffer device;
```
Comment (jserv) [inline README.md:41]: Then, we need to extend `vcam-util` to retrieve the device node of controlling framebuffer.

File: README.md [hunk 22]
```diff
@@ -55,7 +56,7 @@ $ sudo ./vcam-util -l
 You should get:
 ```
 Available virtual V4L2 compatible devices:
-1. vcamfb0(640,480,rgb24) -> /dev/video0
+1. vcamfb0(640,480,rgb24) and fb1(640,480)-> /dev/video0
```
Comment (jserv) [inline README.md:59]: New question raised by this pull request: is it necessary to keep `vcamfb`?

File: vcam-util.c [hunk 23]
```diff
@@ -37,6 +37,7 @@ struct vcam_device_spec device_template = {
     .pix_fmt = VCAM_PIXFMT_RGB24,
     .video_node = "",
     .fb_node = "",
+    .fb_node1 = "",
```
Comment (jserv) [inline vcam-util.c:40]: Ditto. Think of the necessity for yet another framebuffer device node.

**Review by jserv — CHANGES_REQUESTED** [review]
Change the subject to reflect what you have been modifying. It is intended to rework.

File: device.h [hunk 24]
```diff
@@ -72,6 +72,10 @@ struct vcam_device {
     struct proc_dir_entry *vcam_fb_procf;
     struct mutex vcam_mutex;
 
+    /* framebuffer vcamfb_info  */
+    void *fb_priv;
+    char fb_id[FB_NAME_MAXLENGTH];
```
Comment (jserv) [inline device.h:77]: Can you move `fb_id` into the `fb_priv`?

File: fb.h [hunk 25]
```diff
@@ -1,10 +1,25 @@
...
 
+#include <linux/fb.h>
+
 #include "device.h"
 
 struct proc_dir_entry *init_framebuffer(const char *proc_fname,
                                         struct vcam_device *dev);
 void destroy_framebuffer(const char *proc_fname);
 
+int init_vcamfb(struct vcam_device *dev);
```
Comment (jserv) [inline fb.h:12]: You have to distinguish from `init_framebuffer` and `init_vcamfb`. I would like to simplify the interface.

File: fb.h [hunk 26]
```diff
@@ -1,10 +1,25 @@
...
+
+void destroy_vcamfb(struct vcam_device *dev);
+
+void vcam_update_vcamfb(struct vcam_device *dev);
+
+struct vcamfb_info {
```
Comment (jserv) [inline fb.h:18]: Don't do this! `struct vcamfb_info` would be implementation specific. Never make it public. Instead, you can expose some getter function(s) for retrieving.

File: control.c [hunk 27]
```diff
@@ -64,6 +65,7 @@ static ssize_t control_write(struct file *file,
 static int control_iocontrol_get_device(struct vcam_device_spec *dev_spec)
 {
     struct vcam_device *dev;
+    struct vcamfb_info *fb_data;
```
Comment (jserv) [inline control.c:68]: Don't do this. Never expose the implementation details.

File: control.c [hunk 28]
```diff
@@ -74,7 +76,8 @@ static int control_iocontrol_get_device(struct vcam_device_spec *dev_spec)
         dev_spec->pix_fmt = VCAM_PIXFMT_YUYV;
     else
         dev_spec->pix_fmt = VCAM_PIXFMT_RGB24;
-    strncpy((char *) &dev_spec->fb_node, (const char *) &dev->vcam_fb_fname,
+    fb_data = dev->fb_priv;
+    strncpy((char *) &dev_spec->fb_node, (const char *) &fb_data->fb_name,
```
Comment (jserv) [inline control.c:80]: Use a dedicated function to get framebuffer information.

File: README.md (hunk 21 above)
Comment (jserv) [reply README.md:41]: At present, `vcam-util -l` dumps all information, but it would be great if a command option such as `--fbdev` shows the available name of controlling framebuffer. Then, we can program with the device.

File: fb.h [hunk 29]
```diff
@@ -1,10 +1,25 @@
...
+    struct fb_info info;
+    void *addr;
+    unsigned int offset;
+    char fb_name[FB_NAME_MAXLENGTH];
```
Comment (jserv) [inline fb.h:22]: Once this structure is only visible to implementation, it does not make sense to use `fb_name`. Use plain `name` for short.

File: fb.h (hunk 25 above)
Comment (jserv) [reply fb.h:6]: We can now refactor the name scheme of helper functions and make them more consistent. That is,
* `vcamfb_init`
* `vcamfb_destroy`
* `vcamfb_update`
* `vcamfb_get_name`

File: device.c [hunk 30]
```diff
@@ -942,8 +939,8 @@ void destroy_vcam_device(struct vcam_device *vcam)
 
     if (vcam->sub_thr_id)
         kthread_stop(vcam->sub_thr_id);
-    vcam_in_queue_destroy(&vcam->in_queue);
-    destroy_framebuffer(vcam->vcam_fb_fname);
+    vcamfb_destroy(vcam);
+    vfree(vcam->fb_priv);
```
Comment (jserv) [inline device.c:943]: Can you combine the two operations into single function? That is, we don't have to invoke the sequence:
```c
vcamfb_destroy(vcam);
vfree(vcam->fb_priv);
```

Instead, single `vcamfb_destroy(vcam)` can release the allocated resources without explicitly passing `vcam->fb_priv`.

File: device.h [hunk 31]
```diff
@@ -72,6 +72,9 @@ struct vcam_device {
     struct proc_dir_entry *vcam_fb_procf;
     struct mutex vcam_mutex;
 
+    /* framebuffer vcamfb_info  */
```
Comment (jserv) [inline device.h:75]: Change the comment to
```c
/* framebuffer private data */
```

File: README.md (hunk 21 above)
Comment (WayneLin1992) [reply README.md:41]: It's that means  `vcam-util -fbdev` showing framebuffer informations ?

Comment (jserv) [reply README.md:41]: Yes, I meant the new option which outputs the framebuffer device name can help further applications for performing configurations. However, you can implement such feature in next pull request.
